src/Adapter/CartRule/Repository/CartRuleRepository.php

```diff
@@ -328,6 +329,77 @@ public function restrictCartRules(CartRuleId $cartRuleId, array $restrictedCartR
         );
     }
 
+    /**
+     * @param CartRuleId $cartRuleId
+     * @param CarrierId[] $restrictedCarrierIds
+     *
+     * @return void
+     */
+    public function setCarrierRestrictions(CartRuleId $cartRuleId, array $restrictedCarrierIds): void
+    {
+        $this->removeCarrierRestrictions($cartRuleId);
+
+        if (empty($restrictedCarrierIds)) {
+            return;
+        }
+
+        $checkedIds = [];
+        $insertValues = [];
+        foreach ($restrictedCarrierIds as $restrictedCarrierId) {
+            $restrictedCarrierIdValue = $restrictedCarrierId->getValue();
+            // skip duplicate ids if for some reason they exist
+            if (in_array($restrictedCarrierIdValue, $checkedIds, true)) {
+                continue;
+            }
+            $insertValues[] = sprintf('(%d,%d)', $cartRuleId->getValue(), $restrictedCarrierIdValue);
+            $checkedIds[] = $restrictedCarrierIdValue;
+        }
+
+        $this->connection->executeStatement(
+            sprintf(
+                'INSERT INTO %s (`id_cart_rule`, `id_carrier`) VALUES %s',
+                $this->dbPrefix . 'cart_rule_carrier',
+                implode(',', $insertValues)
+            )
+        );
+    }
+
+    /**
+     * @param CartRuleId $cartRuleId
+     *
+     * @return int[]
+     */
+    public function getRestrictedCarrierIds(CartRuleId $cartRuleId): array
+    {
+        $cartRuleIdValue = $cartRuleId->getValue();
+        $results = $this->connection->createQueryBuilder()
+            ->select('crc.id_carrier')
+            ->from($this->dbPrefix . 'cart_rule_carrier', 'crc')
+            ->where('crc.id_cart_rule = :cartRuleId')
+            ->setParameter('cartRuleId', $cartRuleIdValue)
+            ->execute()
+            ->fetchAllAssociative()
+        ;
+
+        if (empty($results)) {
+            return [];
+        }
+
+        return array_map(static function (array $result): int {
+            return (int) $result['id_carrier'];
+        }, $results);
```

Comment:
```suggestion
        return array_map(
            static fn (array $result): int => (int) $result['id_carrier'], 
            $results
          );
```

Not blocking, just for the beauty of art :)